File: public/tools/elements/streams/GtStreams.ts
import {Element, Inject, On, Property} from "../../polymer/Annotations";
import {TabsGenerator, View} from "../../client/router/View";
import {StreamsPlayer} from "./StreamsPlayer";
import {GtAlert, GtBox} from "../widgets/GtBox";
import {StreamsViewers} from "./StreamsViewers";
import {StreamsItem} from "./StreamsItem";
import {PolymerElement} from "../../polymer/PolymerElement";
import {StreamsService, ActiveStream} from "../../services/stream/StreamService";
import {GtStreamsSettings} from "./GtStreamsSettings";
import {GtStreamsWhitelist} from "./GtStreamsWhitelist";

export const StreamsTabs: TabsGenerator = (view, path, user) => [
	{title: "Live Streams", link: "/streams", active: view == GtStreams},
	{title: "Publish", link: "/streams/settings", active: view == GtStreamsSettings},
	{title: "Whitelist", link: "/streams/whitelist", active: view == GtStreamsWhitelist, hidden: !user.promoted}
];

///////////////////////////////////////////////////////////////////////////////
// <gt-streams>

@View("streams", StreamsTabs)
@Element({
	selector: "gt-streams",
	template: "/assets/views/streams.html",
	dependencies: [GtBox, StreamsItem, StreamsViewers, GtAlert, StreamsPlayer]
})
export class GtStreams extends PolymerElement {
	@Inject
	@On({"offline": "StreamOffline"})
	private service: StreamsService;

	@Property
	public selected: ActiveStream = null;

	private StreamOffline(user: number) {
		if (this.selected && this.selected.user == user) {
			this.selected = null;
		}
	}

	private SelectStream(ev: PolymerModelEvent<ActiveStream>) {
		let stream = ev.model.item;
		this.selected = (this.selected && this.selected.user == stream.user) ? null : stream;
	}

	private StreamIsActive(user: number) {
		return this.selected && this.selected.user == user;
	}
}

File: client/src/gt/components/streams/GtStreams.scala
package gt.components.streams

import gt.components.widget.{GtAlert, GtBox}
import gt.components.{GtHandler, Tab, View}
import gt.services.StreamService
import rx.Var
import utils.jsannotation.js
import xuen.Component

/**
  * The composer view.
  */
object GtStreams extends Component[GtStreams](
	selector = "gt-streams",
	templateUrl = "/assets/imports/views/streams.html",
	dependencies = Seq(StreamItem, StreamViewers, StreamPlayer, GtBox, GtAlert)
) with View.Sticky {
	val module = "streams"

	val tabs: TabGenerator = (selector, path, user) => Seq(
		Tab("Streams", "/streams", true)
	)
}

@js class GtStreams extends GtHandler {
	val stream = service(StreamService)
	val list = stream.streams.values ~ (_.map(_.!))
	def hasStreams = list.nonEmpty
	val selected = Var(0)
	def selectStream(user: Int) = selected ~= (c => if (c == user) 0 else user)
}
